crates/rattler_repodata_gateway/src/sparse/mod.rs

```diff
@@ -197,14 +197,81 @@ impl SparseRepoData {
     /// This works by iterating over all elements in the `packages` and
     /// `conda_packages` fields of the repodata and returning the unique
     /// package names.
-    pub fn package_names(&self) -> impl Iterator<Item = &'_ str> {
+    pub fn package_names(
+        &self,
+        package_format_selection: PackageFormatSelection,
+    ) -> impl Iterator<Item = &'_ str> {
         let repo_data = self.inner.borrow_repo_data();
-        let tar_baz2_packages = repo_data.packages.iter().map(|(name, _)| name.package);
-        let conda_packages = repo_data
-            .conda_packages
-            .iter()
-            .map(|(name, _)| name.package);
-        tar_baz2_packages.merge(conda_packages).dedup()
+
+        fn select_package_name<'i>((filename, _): &(PackageFilename<'i>, &'i RawValue)) -> &'i str {
+            filename.package
+        }
+
+        let tar_baz2_packages = repo_data.packages.iter().map(select_package_name);
+        let conda_packages = repo_data.conda_packages.iter().map(select_package_name);
+
+        match package_format_selection {
+            PackageFormatSelection::Both | PackageFormatSelection::PreferConda => {
+                itertools::Either::Left(tar_baz2_packages.merge(conda_packages).dedup())
+            }
+            PackageFormatSelection::OnlyTarBz2 => {
+                itertools::Either::Right(tar_baz2_packages.dedup())
+            }
+            PackageFormatSelection::OnlyConda => itertools::Either::Right(conda_packages.dedup()),
+        }
+    }
+
+    /// Returns the number of packages in this instance.
+    pub fn package_count(&self, package_format_selection: PackageFormatSelection) -> usize {
```

Comment:
I wonder if this is confusing. I think I initially thought it would be a count of package names.

Maybe package_record_count would be more explicit?

Reply:
Or maybe just `record_count`?